### PR TITLE
Viewtiverse: remove context dependencies

### PR DESCRIPTION
**Pull request by Gorash**
In this viewtiverse, the heroes remove the context dependencies for
`get_views`, from the views and python fields (such as domain). To reduce
inconsistencies and the number of rpc.

Current issues:
* There may be inconsistencies in views at the JavaScript level. Some
overrides modify the behavior of get_views or domains on fields via
context keys, therefore by changing the action, the rendering may be
different. However, these views are cached. However, the cache key
(Javascript) does not reflect the entire context, and requires additional
post-processing from the server.
* Multiple rpc for the same rendering. get_views being dependent on the
context, as soon as it changes, a new rpc is performed. In most cases,
when JavaScript needs the same view, there is no change depending on the
context, the rpc is useless.
* Inconsistency when rendering subviews, some views could be different
depending on the context, this context can be modified in the view itself
via the context attributes. However, the JavaScript client does not redo
an rpc for each change of these sub-contexts. Therefore the result may be
inconsistent.

Solution:
Limit as much as possible the number of context keys provided when calling
get_views, and use the context provided as a cache key. The authorized
keys are 'lang' and '*_view_ref'. For the cache key, options are added in
the get_views method.
Instead of using the context, it is inserted into python expressions.
This will be evaluated by JavaScript and thus avoids inconsistencies.

task-3414108
task-3414068

### CODE REVIEW

**Review by aab-odoo — APPROVED** [review]
Ok for the JS part (modulo eslint/prettier) :+1:

File: addons/stock/models/product_strategy.py [hunk 1]
```diff
@@ -157,3 +144,21 @@ def _get_putaway_location(self, product, quantity=0, package=None, packaging=Non
                 checked_locations.add(location)
 
         return None
+
+
+class ProductCategory(models.Model):
```
Comment (Whenrow) [inline addons/stock/models/product_strategy.py:149]: There is already the category class in `stock/models/product.py`

https://github.com/odoo/odoo/blob/4584a27c72f61f9e94bc66f2d0643e1d58074579/addons/stock/models/product.py#L993)

**Review by rco-odoo — APPROVED** [review]
Looks good to me.

**Comment by rco-odoo** [comment]
@robodoo r+

**Pull request closed without merging**